src/v/archival/archival_policy.cc

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "archival/archival_policy.h"
+
+#include "storage/disk_log_impl.h"
+#include "storage/segment.h"
+#include "storage/segment_set.h"
+
+namespace archival {
+
+archival_policy_base::archival_policy_base(
+  model::ntp ntp, model::revision_id rev)
+  : _ntp(std::move(ntp))
+  , _rev(rev) {}
+
+const model::ntp& archival_policy_base::get_ntp() const { return _ntp; }
+const model::revision_id& archival_policy_base::get_revision_id() const {
+    return _rev;
+}
+
+class arch_policy_non_compacted final : public archival_policy_base {
+public:
+    arch_policy_non_compacted(
+      const model::ntp& ntp, const model::revision_id& rev)
+      : archival_policy_base(ntp, rev) {}
+
+    std::optional<manifest>
+    make_local_manifest(storage::log_manager& lm) final {
+        manifest tmp(get_ntp(), get_revision_id());
+        std::optional<storage::log> log = lm.get(get_ntp());
+        if (!log) {
+            return std::nullopt;
+        }
+        auto plog = dynamic_cast<storage::disk_log_impl*>(log->get_impl());
+        if (plog == nullptr) {
+            return std::nullopt;
+        }
+        for (const auto& segment : plog->segments()) {
+            bool closed = !segment->has_appender();
+            bool compacted = segment->is_compacted_segment();
+            if (!closed || compacted) {
+                continue;
+            }
+            manifest::segment_meta meta{
+              .is_compacted = segment->is_compacted_segment(),
+              .size_bytes = segment->size_bytes(),
+              .base_offset = segment->offsets().base_offset,
+              .committed_offset = segment->offsets().committed_offset,
+              .is_deleted_locally = false,
+            };
+            auto path = std::filesystem::path(segment->reader().filename());
+            auto seg_name = path.filename().string();
+            // generate segment path
+            tmp.add(segment_name(seg_name), meta);
```

Comment:
I don't think we fully addressed @senior7515's question about building manifests with all the segments at once. IIRC his question was like: if we add some state to each segment, then could we query the segment set for a log to find a segment that needs to be processed without ever needing to build up a data structure containing all segments.

The manfiest looks pretty small, so its probably ok for now as long as we only do it for one ntp at a time. In any case it would be good to track some optimizations and enhancements like this.

Reply:
It is possible to et rid of manifest and return just a list of segments that log_manager has. Then I'll be able to calculate a diff between this list and cached manifest to find new segments that have to be uploaded. But, because the upload process is asynchronous it is possible that some segments will be updated (e.g. compacted) and will no longer match the policy. For this case alone I'm using not only segment names but also a metadata from the manifest. There is a function called 'validate_metadata' that does this.